### PR TITLE
call center UCR queries & app parsing

### PR DESCRIPTION
**Pull request by snopoke**
@czue 
- get data from UCR tables
- refactor queries into query classes
- generate config object from app during sync

Hold off on merge until UCR tables are all built

:fish:

### CODE REVIEW

File: corehq/apps/callcenter/queries.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+        query = select(
+            self.columns(include_type_in_result, distinct_docs=distinct_docs)
+        ).where(and_(
+            *all_where
+        )).group_by(
+            *self.group_by(include_type_in_result)
+        )
+        return query
+
+class CaseQueryOpenedClosed(CaseQuery):
```
Comment (dimagimon) [inline corehq/apps/callcenter/queries.py]: E302 expected 2 blank lines, found 1

File: corehq/apps/callcenter/indicator_sets.py [hunk 2]
```diff
@@ -296,158 +287,60 @@ def _add_case_data_by_type(self, queryset, indicator_prefix, range_name, all_typ
...
-    def add_case_data(self, query_fn, slug, indicator_config, type_column='type', legacy_prefix=None):
-        include_types = indicator_config.all_types or indicator_config.types
-        include_total = indicator_config.total.active
-        limit_types = indicator_config.types
-
-        q_type_exclude = {type_column: self.cc_case_type}
         if include_total and include_types and not limit_types:
-            logger.debug('Adding case_data %s: totals and all types: %s', slug, indicator_config.total.date_ranges)
-            for range_name in indicator_config.total.date_ranges:
+            logger.debug('Adding case_data %s: totals and all types: %s', slug, indicator_config.totals.date_ranges)
```
Comment (dimagimon) [inline corehq/apps/callcenter/indicator_sets.py]: E501 line too long (116 > 115 characters)

Comment (snopoke) [inline corehq/apps/callcenter/indicator_sets.py]: ignoring

File: corehq/sql_db/connections.py [hunk 3]
```diff
@@ -25,6 +27,23 @@ def __init__(self, connection_string):
         # Session is the actual constructor object
         self.Session = scoped_session(self._session_factory)
 
+    @property
+    def session_context(self):
```
Comment (czue) [inline corehq/sql_db/connections.py]: :+1:

File: corehq/apps/callcenter/indicator_sets.py [hunk 4]
```diff
@@ -316,94 +313,171 @@ def _case_query_opened_closed(self, opened_or_closed, values, filters, lower, up
...
         include_total = indicator_config.total.active
         limit_types = indicator_config.types
 
-        q_type_exclude = {type_column: self.cc_case_type}
         if include_total and include_types and not limit_types:
             logger.debug('Adding case_data %s: totals and all types: %s', slug, indicator_config.total.date_ranges)
             for range_name in indicator_config.total.date_ranges:
                 lower, upper = self.date_ranges[range_name]
-                results = query_fn(['case_owner', 'type'], ~Q(**q_type_exclude), lower, upper)
+                results = query_fn(True, None, lower, upper)
```
Comment (czue) [inline corehq/apps/callcenter/indicator_sets.py]: nit: might be easier to understand this if these calls used kwargs

File: corehq/apps/callcenter/queries.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+        ]).where(and_(
+            operators.ge(self.sql_table.c.time_end, start_date),
+            operators.lt(self.sql_table.c.time_end, end_date),
+            operators.in_op(self.sql_table.c.user_id, self.users_needing_data),
+            self.sql_table.c.xmlns == xmlns,
+        )).group_by(
+            self.sql_table.c.user_id
+        )
+
+        return self._run_query(query)
```
Comment (czue) [inline corehq/apps/callcenter/queries.py]: :+1: this commit makes things much more understandable

File: corehq/apps/callcenter/app_parser.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        period = split[-1]
+        slug = '{}_{}'.format(*split[0:2])
+        type = None
+        if len(split) == 4:
+            type = split[-2]
+
+        return ParsedIndicator(slug, type, period, False)
+
+
+def get_indicators_used_in_app(app):
```
Comment (czue) [inline corehq/apps/callcenter/app_parser.py]: do we also need to check for indicators in forms?

File: corehq/apps/callcenter/models.py [hunk 7]
```diff
@@ -1,85 +1,81 @@
...
-            cases_closed=default_typed(),
-        )
-
-    def save(self, **params):
-        super(CallCenterIndicatorConfig, self).save(**params)
-        CallCenterIndicatorConfig.for_domain.clear(self.__class__, self.domain)
+            return indicator
+
+
+class CallCenterIndicatorConfig(JsonObject):
```
Comment (czue) [inline corehq/apps/callcenter/models.py]: was `domain` intentionally removed?

File: corehq/apps/callcenter/models.py [hunk 8]
```diff
@@ -1,85 +1,111 @@
...
-    cases_active = SchemaProperty(ByTypeIndicator)
-    cases_opened = SchemaProperty(ByTypeIndicator)
-    cases_closed = SchemaProperty(ByTypeIndicator)
 
-    @classmethod
-    @skippable_quickcache(['domain'], lambda *_: settings.UNIT_TESTING)
-    def for_domain(cls, domain):
-        res = get_docs_in_domain_by_class(domain, cls)
-        return res[0] if res else cls.default_config(domain)
+class CallCenterIndicatorConfig(JsonObject):
```
Comment (czue) [inline corehq/apps/callcenter/models.py]: copying comment here to show up in diff: was `domain` intentionally removed?

**Comment by czue** [comment]
wow, you went on a bit of a rampage. this is looking good - I didn't review the query logic too closely but I assume the tests cover that?

File: corehq/apps/callcenter/app_parser.py (hunk 6 above)
Comment (snopoke) [inline corehq/apps/callcenter/app_parser.py]: ah, yes we do. Thanks.

File: corehq/apps/callcenter/models.py (hunk 8 above)
Comment (snopoke) [inline corehq/apps/callcenter/models.py]: Since this never gets saved there is no need to have domain in it.

**Comment by snopoke** [comment]
Yea, I figured might as well get it done while I had my head in the space

**Comment by snopoke** [comment]
Yes, tests should cover the queries

File: corehq/apps/callcenter/tests/test_use_fixtures_configuration.py [hunk 9]
```diff
@@ -1,6 +1,15 @@
+import os
+
 from django.test import SimpleTestCase
+
+from corehq.apps.callcenter.app_parser import get_indicators_used_in_app, parse_indicator, ParsedIndicator, \
```
Comment (dimagimon) [inline corehq/apps/callcenter/tests/test_use_fixtures_configuration.py]: F401 'get_indicators_used_in_app' imported but unused

File: corehq/apps/callcenter/app_parser.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+            instance_id = instance_node.attrib['id']
+            search_term = "instance('{}')".format(instance_id)
+            for node in wrapped_form.model_node.xpath('*[contains(@calculate, "{}")]'.format(search_term)):
+                xpath = node.attrib['calculate']
+                indicator_name = xpath.split('/')[-1]
+                indicators.add(indicator_name)
+
+    return indicators
+
+def _get_indicators_in_detail(detail):
```
Comment (dimagimon) [inline corehq/apps/callcenter/app_parser.py]: E302 expected 2 blank lines, found 1

File: corehq/apps/callcenter/models.py [hunk 11]
```diff
@@ -1,85 +1,121 @@
...
-        CallCenterIndicatorConfig.for_domain.clear(self.__class__, self.domain)
+        config.legacy_forms_submitted = include_legacy
+        config.legacy_cases_total = include_legacy
+        config.legacy_cases_active = include_legacy
+
+        return config
+
+    def set_indicator(self, parsed_indicator):
+        if parsed_indicator.is_legacy:
+            legacy_indicator = setattr(self, 'legacy_{}'.format(parsed_indicator.category), True)
```
Comment (dimagimon) [inline corehq/apps/callcenter/models.py]: F841 local variable 'legacy_indicator' is assigned to but never used

File: corehq/apps/callcenter/indicator_sets.py [hunk 12]
```diff
@@ -296,158 +287,66 @@ def _add_case_data_by_type(self, queryset, indicator_prefix, range_name, all_typ
...
-    def add_case_data(self, query_fn, slug, indicator_config, type_column='type', legacy_prefix=None):
-        include_types = indicator_config.all_types or indicator_config.types
-        include_total = indicator_config.total.active
-        limit_types = indicator_config.types
-
-        q_type_exclude = {type_column: self.cc_case_type}
         if include_total and include_types and not limit_types:
-            logger.debug('Adding case_data %s: totals and all types: %s', slug, indicator_config.total.date_ranges)
-            for range_name in indicator_config.total.date_ranges:
+            logger.debug('Adding case_data %s: totals and all types: %s', slug, indicator_config.totals.date_ranges)
```
Comment (dimagimon) [inline corehq/apps/callcenter/indicator_sets.py]: E501 line too long (116 > 115 characters)

Comment (dimagimon) [inline corehq/apps/callcenter/indicator_sets.py]: E501 line too long (116 > 115 characters)

Comment (dimagimon) [inline corehq/apps/callcenter/indicator_sets.py]: E501 line too long (116 > 115 characters)

File: corehq/apps/callcenter/indicator_sets.py [hunk 13]
```diff
@@ -297,158 +287,66 @@ def _add_case_data_by_type(self, queryset, indicator_prefix, range_name, all_typ
...
```
Comment (dimagimon) [inline corehq/apps/callcenter/indicator_sets.py]: E501 line too long (116 > 115 characters)

File: corehq/apps/callcenter/indicator_sets.py [hunk 14]
```diff
@@ -297,158 +287,66 @@ def _add_case_data_by_type(self, queryset, indicator_prefix, range_name, all_typ
...
```
Comment (dimagimon) [inline corehq/apps/callcenter/indicator_sets.py]: E501 line too long (116 > 115 characters)

Comment (dimagimon) [inline corehq/apps/callcenter/indicator_sets.py]: E501 line too long (116 > 115 characters)

Comment (dimagimon) [inline corehq/apps/callcenter/indicator_sets.py]: E501 line too long (116 > 115 characters)

**Comment by snopoke** [comment]
@czue I think we're finally ready for this to be merged (once :white_check_mark: )

**Comment by snopoke** [comment]
:white_check_mark:

File: corehq/apps/callcenter/fixturegenerators.py [hunk 15]
```diff
@@ -59,6 +60,13 @@ def __call__(self, user, version, last_sync=None, app=None):
                     'app_id': app.get_id
                 })
 
+        if config:
+            _assert = soft_assert(['skelly_at_dimagi_dot_com'.replace('_at_', '@').replace('_dot_', '.')])
```
Comment (czue) [inline corehq/apps/callcenter/fixturegenerators.py]: getting fancy with this

File: corehq/apps/callcenter/app_parser.py [hunk 16]
```diff
@@ -49,6 +53,24 @@ def get_indicators_used_in_app(app):
...
+    for form in app.get_forms(bare=True):
+        wrapped_form = form.wrapped_xform()
+        if not wrapped_form.exists():
+            continue
+
+        instance_node = wrapped_form.model_node.find('{f}instance[@src="jr://fixture/indicators:call-center"]')
+        if instance_node.exists():
+            instance_id = instance_node.attrib['id']
+            search_term = "instance('{}')".format(instance_id)
+            for node in wrapped_form.model_node.xpath('*[contains(@calculate, "{}")]'.format(search_term)):
```
Comment (czue) [inline corehq/apps/callcenter/app_parser.py]: not 100% sure i'm reading this right, but is this making the assumption that these only appear in calculates? (seems probably ok, though i can imagine some rare/advanced forms maybe using it elsewhere)

**Comment by czue** [comment]
no blockers, merging. :phone: :rocket:

**Pull request merged**